README.md
# ReInk
An utility to reset Epson printer ink counters.

## About

ReInk - is an **experimental** utility to reset ink level in new chiped *Epson*
 cartridges programmatically through regular printer interface (i.e. without
 any additional hardware controllers).

This utility is supposed to be an Open Source alternative to Windows program
[SSC Service Utility](http://www.ssclg.com/epsone.shtml)

## Disclaimer

THIS UTILITY MAY DAMAGE YOUR PRINTER !!!

USE CAREFULLY AND AT YOUR OWN RISK.

## Features
To find out features of current ReInk version look at the source code. :)

Major functions list:
* Infrastructure for communicating with printer by means of IEEE 1284.4 protocol (initializing, opening/closing
channels, service discovery, data transmission and so on).
* Read from arbitary EEPROM address for supported printers.
* Write to arbitary EEPROM address for supported printers.
* Get ink levels for every EPSON Stylus printer I think (almost "copied" from [escputil](http://gimp-print.sourceforge.net/)).
* Reset ink levels for supported printers.
* Reset waste ink counter for supported printers.
* Can generate test reports, containing information about printer.

## Supported printers
* EPSON Stylus Photo 790
* EPSON Stylus Color 580
* EPSON Stylus Photo 1290
* EPSON Stylus Color 680
* EPSON Stylus Photo T50
* EPSON Stylus Photo P50

## How to get my printer supported?
See [here](http://lion-simba.github.io/reink/report.html).

## Usage
Compile (`make`) and run (`./reink`) the program without arguments. :)
In most cases you would need to power-off and then power-on your printer after reseting ink level. This will allow printer
to save new data in the cartridges.


debian/changelog
reink (0.6) UNRELEASED; urgency=low

  * Support for new printers:
  *   Epson Stylus Photo P50
  *   Epson Stylus Photo T50

 -- Alexey Osipov <public@alexey.osipov.name>  Sun, 17 Jan 2016 18:03:40 +0600

reink (0.5) UNRELEASED; urgency=low

  * Support for new printer: Epson Stylus Color 680.

 -- Alexey Osipov <public@alexey.osipov.name>  Sun, 27 Sep 2009 12:17:44 +0700

reink (0.4) UNRELEASED; urgency=low

  * Storing printers data in separate source file - printers.c.
  * Epson Stylus Color 580 should really work now.
  * Support for new printer: Epson Stylus Photo 1290.

 -- Alexey Osipov <public@alexey.osipov.name>  Sun, 27 Sep 2009 00:36:39 +0700

reink (0.3) UNRELEASED; urgency=low

  * Support for new printer: Epson Stylus Color 580.
  * Ability to reset printer's waste ink counter (-s switch).

 -- Alexey Osipov <public@alexey.osipov.name>  Fri, 25 Sep 2009 21:13:20 +0700

reink (0.2.2) UNRELEASED; urgency=low

  * Flush stdout before starting bruteforce.
  * Ability to enforce full debug mode while making testreport.

 -- Alexey Osipov <public@alexey.osipov.name>  Wed, 15 Jul 2009 20:57:03 +0700

reink (0.2.1) UNRELEASED; urgency=low

  * Correct read_eeprom_address in case of two-byte addresses.

 -- Alexey Osipov <public@alexey.osipov.name>  Tue, 02 Jun 2009 15:32:06 +0700

reink (0.2) UNRELEASED; urgency=low

  * Refactoring of command building facility. Now commands build by its
    known format and not by "magic".
  * Support for two-bytes eeprom_read and eeprom_write commands.
  * Imporved test report facility: now capable to guess the secret model code
    (with brute force); two-pass test report for help me to guess eeprom
    addresses with ink data.

 -- Alexey Osipov <public@alexey.osipov.name>  Sat, 30 May 2009 14:07:55 +0700

reink (0.1) UNRELEASED; urgency=low

  * Initial release. (Closes: #XXXXXX)

 -- Alexey Osipov <public@alexey.osipov.name>  Sat, 30 May 2009 13:58:45 +0700


d4lib.h
/*
 *
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 2 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 59 Temple Place - Suite 330, Boston, MA 02111-1307, USA.
 */

#ifndef D4LIB_H

#define D4LIB_H

extern int debugD4;   /* allow printout of debug informations */

extern int EnterIEEE(int fd);
extern int Init(int fd);
extern int Exit(int fd);
extern int GetSocketID(int fd, const char *serviceName);
extern int OpenChannel(int fd, unsigned char sockId, int *sndSz, int *rcvSz);
extern int CloseChannel(int fd, unsigned char socketID);
extern int CreditRequest(int fd, unsigned char socketID);
extern int Credit(int fd, unsigned char socketID, int credit);

/* convenience function */
extern int SafeWrite(int fd, const void *data, int len);
extern int askForCredit(int fd, unsigned char socketID, int *sndSz, int *rcvSz);
extern int writeData(int fd, unsigned char socketID, const unsigned char *buf, int len, int eoj);
extern int readData(int fd, unsigned char socketID, unsigned char *buf, int len);
extern int readAnswer(int fd, unsigned char *buf, int len);
extern void flushData(int fd, unsigned char socketID);
extern void clearSndBuf(int fd);
extern void setDebug(int debug);

extern int d4WrTimeout;
extern int d4RdTimeout;
extern int ppid;

#if D4_DEBUG
#define DEBUG 1
#endif

#endif



printers.h
/* This file is part of ReInk.
 * Copyright (C) 2008-2016 Alexey Osipov public@alexey.osipov.name
 *
 * ReInk is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 3 of the License, or
 * (at your option) any later version.
 *
 * ReInk is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with ReInk. If not, see <http://www.gnu.org/licenses/>.
 */
 
 
//printer models 
#define PM_UNKNOWN		0	//Unknown model
#define PM_SP790		1	//EPSON Stylus Photo 790
#define PM_SC580		2	//EPSON Stylus Color 580
#define PM_SP1290		3	//EPSON Stylus Photo 1290
#define PM_SC680		4	//EPSON Stylus Color 680
#define PM_SPT50		5	//EPSON Stylus Photo T50
#define PM_SPP50		6	//EPSON Stylus Photo P50
#define PM_SP1410		7	//EPSON Stylus Photo 1410

//ink counters EEPROM map
#define INK_BLACK			0x01
#define INK_CYAN			0x02
#define INK_MAGENTA			0x04
#define INK_YELLOW			0x08
#define INK_LIGHTCYAN		0x10
#define INK_LIGHTMAGENTA	0x20
typedef struct _ink_map {
	unsigned int  mask; //bit sum of INK_ defines of available inks	
	unsigned char black[4];
	unsigned char cyan[4];
	unsigned char magenta[4];
	unsigned char yellow[4];
	unsigned char lightcyan[4];
	unsigned char lightmagenta[4];
} ink_map_t;

//waste counter EEPROM map
typedef struct _waste_map {
	unsigned char len;		//count of used bytes from addr
	unsigned char addr[4];
} waste_map_t;

//the printer
#define MAX_NAME_LEN	100
#define MAX_MODEL_LEN	100
typedef struct _printer_s {
	unsigned char name[MAX_NAME_LEN];			//printer textual name
	unsigned char model_name[MAX_MODEL_LEN];	//printer model name as returned by printer itself
	unsigned char model_code[2];	//"password" for this printer
	int twobyte_addresses;			//is printer's EEPROM uses two-byte addresses?
	ink_map_t inkmap;
	waste_map_t wastemap;
} printer_t;

extern printer_t printers[];
extern const unsigned int printers_count;


printers.c
/* This file is part of ReInk.
 * Copyright (C) 2008-2016 Alexey Osipov public@alexey.osipov.name
 * Copyright (C) 2014 Andrei Komarovskikh andrei.komarovskikh@gmail.com
 *
 * ReInk is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 3 of the License, or
 * (at your option) any later version.
 *
 * ReInk is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with ReInk. If not, see <http://www.gnu.org/licenses/>.
 */

#include "printers.h"

printer_t printers[] = {
	[PM_UNKNOWN] = {
		.name = "Unknown printer",
		.model_name = "Unknown printer",
		.model_code = {0x00, 0x00},
		.twobyte_addresses = 0,
		.inkmap = {
			.mask = 			0,
			.black = 			{0x00, 0x00, 0x00, 0x00},
			.cyan = 			{0x00, 0x00, 0x00, 0x00},
			.magenta = 			{0x00, 0x00, 0x00, 0x00},
			.yellow = 			{0x00, 0x00, 0x00, 0x00},
			.lightcyan = 		{0x00, 0x00, 0x00, 0x00},
			.lightmagenta = 	{0x00, 0x00, 0x00, 0x00},
		},
		.wastemap = {
			.len = 				0,
			.addr = 			{0x00, 0x00, 0x00, 0x00},
		},
	},
	[PM_SP790] = {
		.name = "EPSON Stylus Photo 790",
		.model_name = "Stylus Photo 790",
		.model_code = {0x06, 0x31},
		.twobyte_addresses = 0,
		.inkmap = {
			.mask = INK_BLACK | INK_CYAN | INK_MAGENTA | INK_YELLOW | INK_LIGHTCYAN | INK_LIGHTMAGENTA,
			.black = 			{0x02, 0x03, 0x04, 0x05},
			.cyan = 			{0x0e, 0x0f, 0x10, 0x11},
			.magenta = 			{0x0a, 0x0b, 0x0c, 0x0d},
			.yellow = 			{0x06, 0x07, 0x08, 0x09},
			.lightcyan = 		{0x16, 0x17, 0x18, 0x19},
			.lightmagenta = 	{0x12, 0x13, 0x14, 0x15},
		},
		.wastemap = {
			.len = 				2,
			.addr = 			{0x1a, 0x1b, 0x00, 0x00},
		},
	},
	[PM_SC580] = {
		.name = "EPSON Stylus Color 580",
		.model_name = "Stylus COLOR 580",
		.model_code = {0x06, 0x1b},
		.twobyte_addresses = 0,
		.inkmap = {
			.mask = INK_BLACK | INK_CYAN | INK_MAGENTA | INK_YELLOW,
			.black = 			{0x44, 0x45, 0x46, 0x47},
			.cyan = 			{0x52, 0x53, 0x54, 0x55},
			.magenta = 			{0x4c, 0x4d, 0x4e, 0x4f},
			.yellow = 			{0x48, 0x49, 0x4a, 0x4b},
			.lightcyan = 		{0x00, 0x00, 0x00, 0x00},
			.lightmagenta = 	{0x00, 0x00, 0x00, 0x00},
		},
		.wastemap = {
			.len = 				2,
			.addr = 			{0x66, 0x67, 0x00, 0x00},
		},
	},
	[PM_SP1290] = {
		.name = "EPSON Stylus Photo 1290",
		.model_name = "Stylus Photo 1290",
		.model_code = {0x07, 0x19},
		.twobyte_addresses = 0,
		.inkmap = {
			.mask = INK_BLACK | INK_CYAN | INK_MAGENTA | INK_YELLOW | INK_LIGHTCYAN | INK_LIGHTMAGENTA,
			.black = 			{0x44, 0x45, 0x46, 0x47},
			.cyan = 			{0x4c, 0x4d, 0x4e, 0x4f},
			.magenta = 			{0x50, 0x51, 0x52, 0x53},
			.yellow = 			{0x48, 0x49, 0x4a, 0x4b},
			.lightcyan = 		{0x54, 0x55, 0x56, 0x57},
			.lightmagenta = 	{0x58, 0x59, 0x5a, 0x5b},
		},
		.wastemap = {
			.len = 				4,
			.addr = 			{0x6c, 0x6d, 0x6e, 0x6f},
		},
	},
	[PM_SC680] = {
		.name = "EPSON Stylus Color 680",
		.model_name = "Stylus COLOR 680",
		.model_code = {0x06, 0x15},
		.twobyte_addresses = 0,
		.inkmap = {
			.mask = INK_BLACK | INK_CYAN | INK_MAGENTA | INK_YELLOW | INK_LIGHTCYAN | INK_LIGHTMAGENTA,
			.black = 			{0x02, 0x03, 0x04, 0x05},
			.cyan = 			{0x0e, 0x0f, 0x10, 0x11},
			.magenta = 			{0x0a, 0x0b, 0x0c, 0x0d},
			.yellow = 			{0x06, 0x07, 0x08, 0x09},
			.lightcyan = 		{0x16, 0x17, 0x18, 0x19},
			.lightmagenta = 	{0x12, 0x13, 0x14, 0x15},
		},
		.wastemap = {
			.len = 				2,
			.addr = 			{0x1a, 0x1b, 0x00, 0x00},
		},
	},
	[PM_SPT50] = {
		.name = "Epson Stylus Photo T50",
		.model_name = "Epson Stylus Photo T50",
		.model_code = {0x77, 0x00},
		.twobyte_addresses = 0,
		.inkmap = {
			.mask = INK_BLACK | INK_CYAN | INK_MAGENTA | INK_YELLOW | INK_LIGHTCYAN | INK_LIGHTMAGENTA,
			.black = 			{0x02, 0x03, 0x04, 0x05},
			.cyan = 			{0x0e, 0x0f, 0x10, 0x11},
			.magenta = 			{0x0a, 0x0b, 0x0c, 0x0d},
			.yellow = 			{0x06, 0x07, 0x08, 0x09},
			.lightcyan = 		{0x16, 0x17, 0x18, 0x19},
			.lightmagenta = 	{0x12, 0x13, 0x14, 0x15},
		},
		.wastemap = {
			.len = 				4,
			.addr = 			{0x1c, 0x1d, 0x1e, 0x1f},
		},
	},
	[PM_SPP50] = {
		.name = "Epson Stylus Photo P50",
		.model_name = "Epson Stylus Photo P50",
		.model_code = {0x77, 0x00},
		.twobyte_addresses = 0,
		.inkmap = {
			.mask = INK_BLACK | INK_CYAN | INK_MAGENTA | INK_YELLOW | INK_LIGHTCYAN | INK_LIGHTMAGENTA,
			.black = 			{0x02, 0x03, 0x04, 0x05},
			.cyan = 			{0x0e, 0x0f, 0x10, 0x11},
			.magenta = 			{0x0a, 0x0b, 0x0c, 0x0d},
			.yellow = 			{0x06, 0x07, 0x08, 0x09},
			.lightcyan = 		{0x16, 0x17, 0x18, 0x19},
			.lightmagenta = 	{0x12, 0x13, 0x14, 0x15},
		},
		.wastemap = {
			.len = 				4,
			.addr = 			{0x1c, 0x1d, 0x1e, 0x1f},
		},
	},
	[PM_SP1410] = {
		.name = "EPSON Stylus Photo 1410",
		.model_name = "Stylus Photo 1410",
		.model_code = {0x10, 0x20},
		.twobyte_addresses = 0,
		.inkmap = {
			.mask = INK_BLACK | INK_CYAN | INK_MAGENTA | INK_YELLOW | INK_LIGHTCYAN | INK_LIGHTMAGENTA,
			.black = 			{0x04, 0x05, 0x06, 0x07},
			.cyan = 			{0x08, 0x09, 0x0a, 0x0b},
			.magenta = 			{0x0c, 0x0d, 0x0e, 0x0f},
			.yellow = 			{0x10, 0x11, 0x12, 0x13},
			.lightcyan = 		{0x14, 0x15, 0x16, 0x17},
			.lightmagenta = 	{0x18, 0x19, 0x1a, 0x1b},
		},
		.wastemap = {
			.len = 				0,
			.addr = 			{0x00, 0x00, 0x00, 0x00},
		},
	}
};

const unsigned int printers_count = sizeof(printers) / sizeof(printers[0]);
